docs/extend/static-code-analysis.md

```diff
@@ -0,0 +1,86 @@
+# Static Code Analysis
+
+Static code analysis is the process of analyzing the source code against a set of rules to find bugs, code smells, and security vulnerabilities. This is a great way to improve the quality of your code and to find potential issues before they are deployed to production. An example is validating the typings of a function to ensure that the function is called with the correct arguments.
+
+Flarum provides a static code analysis package based on [PHPStan](https://phpstan.org/) that can be added to your extension. In this guide, we will show you how to add the package to your extension and how to run the analysis.
+
+## Setup
+
+:::tip [Flarum CLI](https://github.com/flarum/cli)
+
+You can use the CLI to automatically add and update the infrastructure for phpstan to your code:
+
+```bash
+$ flarum-cli infra phpstan
+```
+
+:::
+
+First you need to require the `flarum/phpstan` package in your extension. You can do this by running the following command in the root of our extension:
+
+```bash
+composer require --dev flarum/phpstan:^1.0
+```
+
+Next, you need to create a `phpstan.neon` file in the root of your extension. This file contains [the configuration for PHPStan](https://phpstan.org/config-reference). You can copy the following configuration into the file:
+
+```neon
+includes:
+  - vendor/flarum/phpstan/extension.neon
+
+parameters:
+	# The level will be increased in Flarum 2.0
+	level: 5
+	paths:
+	  - src
+	  - extend.php
+	excludePaths:
+		- *.blade.php
+	checkMissingIterableValueType: false
+	databaseMigrationsPath: ['migrations']
+```
+
+Finally, you need to add the following script to your `composer.json` file:
+
+```json
+{
+  "scripts": {
+    "analyse:phpstan": "phpstan analyse",
+    "clear-cache:phpstan": "phpstan clear-result-cache"
+  },
+  "scripts-descriptions": {
+    "analyse:phpstan": "Run static analysis"
+  }
+}
+```
+
+## Running the analysis
+
+To run the analysis, you can run the following command in the root of your extension:
+
+```bash
+composer analyse:phpstan
+```
+
+If you want to clear the cache before running the analysis, you can run the following command:
+
+```bash
+composer clear-cache:phpstan && composer analyse:phpstan
+```
+
+## GitHub Actions
+
+You can also run the analysis using GitHub Actions. Checkout the page on [GitHub Actions](github-actions.md) for more information.
+
+## Tips
+
+### Extended model attribute types
+
+PHPStan needs to be able to determine the type of an attribute added to an existing model. To do this you can use the `Extend\Model(...)->cast(...)` method.
+
+For example, if your extension were to add a `is_cool` attribute to the `User` model, you can use [attribute casting](https://laravel.com/docs/8.x/eloquent-mutators#attribute-casting) to explicitly define the attribute as boolean. The `flarum/phpstan` package will automatically detect this and communicate it to PHPStan.
+
+```php
+(new Extend\Model(User::class))
+    ->cast('is_cool', 'bool'),
+```
```

Comment:
Just one question about this section. The last error that I couldn't fix is:

    Call to an undefined method Flarum\Discussion\Discussion::sources()

This is a [method I added in extend.php](https://github.com/club-1/flarum-ext-cross-references/blob/2731136be37dcdbabbd41846b6007d2d097a2a25/extend.php#L48-L50) like this:

```php
    (new Extend\Model(Discussion::class))
        ->belongsToMany('sources', Discussion::class, 'discussion_reference', 'target_id', 'source_id')
        ->belongsToMany('targets', Discussion::class, 'discussion_reference', 'source_id', 'target_id'),
```

How can I make PHPStan understand method declared like this?

I thought about using the `cast()` method as it seems it is supposed to solve this kind of issues but it looks like is is only for primitive types.

Reply:
The package should have logic for all relationship methods, I need to check again `belongsToMany` might have been fogotten

Reply:
@n-peugnet can you link to the codebase, `belongsToMany` is correctly picked up by the phpstan package when used on the framework, so not sure why it isn't working on your end.

Reply:
The extension is this one : https://github.com/club-1/flarum-ext-cross-references

I added phpstan setup in this branch : https://github.com/club-1/flarum-ext-cross-references/pull/25

For now it fails with this issue : https://github.com/club-1/flarum-ext-cross-references/actions/runs/4143686829/jobs/7165830767#step:4:15

    Stub file /home/runner/work/flarum-ext-cross-references/flarum-ext-cross-references/vendor/vendor/nunomaduro/larastan/stubs/Enumerable.stub does not exist.

As my commit is not yet available in [flarum/phpstan](https://github.com/flarum/phpstan).

Reply:
@n-peugnet it is now

Reply:
Thanks, I will push a new commit right now.

Reply:
@SychO9 I just updated `flarum/phpstan` to the latest version and now everything is working fine. The stub files are now found as expected and mysteriously, the `Call to an undefined method Flarum\Discussion\Discussion::sources()` error disappeared.

Well, good for us I guess!

Reply:
nice :) thanks for the help!